### ISSUE TITLE
Travis CI linux-aarch64, linux-ppc64le jobs failing

### ISSUE THREAD

**Issue by mfansler** [conda-forge/status#185]
Travis CI **linux-aarch64** and **linux-ppc64le** jobs continue to have high failure rates (10%-25%). This manifests either as infinite queuing or premature cancellation. Restarting the individual build jobs is often sufficient, however, maintainers may also consider moving builds to Azure following https://github.com/conda-forge/status/issues/185#issuecomment-2276925290.

**Comment by mfansler** [comment]
I have been in touch with Travis support via email but no resolution yet.

**Comment by h-vetinari** [comment]
I've seen the issue on the [gtest](https://github.com/conda-forge/gtest-feedstock) feedstock as well, independently of R. In any case, I'm not 100% sure this qualifies as "major". According to the [status page](https://conda-forge.org/status/) we build around 100-150x more on azure than on travis, so <0.5% of our builds are affected[^1], _and_ it's possible (at least in principle) to switch them to azure (either emulated or cross-compiled).

I know this is splitting hairs a bit, so no need to change anything per se (I was thinking along the lines of avoiding a "boy who cried wolf" situation where people evenetually don't take our status seriously, but one time isn't going to do that).

That said, thanks a lot for trying to the bottom of this @mfansler! 🙏 

[^1]: halved from 1% because aarch builds are still working

**Comment by jakirkham** [comment]
Was debating between "degraded" and "major outage". Ok with using "degraded' instead

That said, this appears to be affecting all(?) native `linux_ppc64le` builds and the R migration. So it seemed worthy of it in this case

**Comment by jakirkham** [comment]
Mervin, have you heard anything from Travis CI?

FWIW it seems Travis users outside conda-forge have the same issue. So it is not just us

**Comment by mfansler** [comment]
No word since when I created this. I just sent a ping to see if they have any updates.

**Comment by h-vetinari** [comment]
... https://github.com/conda-forge/conda-forge.github.io/issues/1521 ... 🙄

**Comment by h-vetinari** [comment]
It's been more than a week. Any affected feedstocks should consider either of the following changes in `conda-forge.yml`:
* moving to cross-compilation (might need recipe changes)
  ```yaml
  build_platform:
    linux_ppc64le: linux_64
  ```
* or emulation (much slower, but shouldn't need changes)
  ```yaml
  provider:
    linux_ppc64le: azure
  ```

**Comment by dhirschfeld** [comment]
Do you know of any example PR where a recipe was moved to using cross-compilation for `linux_ppc64le`?

**Comment by h-vetinari** [comment]
You mean for R or in general?

**Comment by hmaarrfk** [comment]
xref: https://github.com/conda/conda-build/issues/5349 (just linking here since I tried to move a package out of PPC64le and hit this)

**Comment by h-vetinari** [comment]
> xref: [conda/conda-build#5349](https://github.com/conda/conda-build/issues/5349) (just linking here since I tried to move a package out of PPC64le and hit this)

That should be a very rare case though. Cross-compilation and `noarch: python` aren't often mixed, because if an output is actually noarch, it suffices to build it just once (e.g. on `linux-64`).

**Comment by dhirschfeld** [comment]
> You mean for R or in general?

In general. The actual feedstock where I'm hitting this is a `go` recipe.

**Comment by mfansler** [comment]
Don't know how much of this will be helpful for other contexts, but here's an example for conversion to cross-compilation on an R feedstock: https://github.com/conda-forge/r-phylobase-feedstock/pull/10

Our recipe (`meta.yaml`) updates include these changes to `build:`:

- adding `cross-r-base`
- adding all required `r-*` dependencies
- all the above go under a `# [build_platform != target_platform]` selector

For `conda-forge.yaml`, we use (as already mentioned):

```yaml
build_platform:
  linux_ppc64le: linux_64
test: native_and_emulated
```

<sub>**NB:** I usually switch `linux_aarch64` to cross-compile as well. If one works they usually both work and the cross-compilation has negligible time difference.</sub>

It is not infrequent that we also need to patch the source's build scripts. Since CRAN native builds everything, our upstreams are not always considering cross-compilation, e.g., they use autoconf scripts that include run tests. Often it can be easiest to simply skip such configure scripts and directly provide pre-determined compilation flags.

**Comment by mfansler** [comment]
> It's been more than a week. Any affected feedstocks should consider either of the following changes in `conda-forge.yml`:
> 
> * moving to cross-compilation (might need recipe changes)
>   ```yaml
>   build_platform:
>     linux_ppc64le: linux_64
>   ```
> * or emulation (much slower, but shouldn't need changes)
>   ```yaml
>   provider:
>     linux_ppc64le: azure
>   ```

Just want to clarify the explicit combinations here:

| `build_platform`              | `provider` | CI - Build Mode |
| :------------  | :------: | :---- |
| linux_ppc64le  |   default   | Travis CI - native |
| linux_ppc64le  |   azure   | Azure - emulate (*slow!*)  |
| linux_64        |   default   | Azure - cross-compile (+ emulated tests)  |
| linux_64        |   azure   | Azure - cross-compile (+ emulated tests)  |

**Comment by minrk** [comment]
I can't find any real competitors to Travis for IBM architectures. But I did find that OSU's Open Source Lab hosts (IBM sponsored) Jenkins instances for [ppc](https://osuosl.org/services/powerdev/) and [s390x](https://osuosl.org/services/ibm-z/) for open source. I'm guessing they are not really prepared to handle conda-forge's scale, but it might be worth a contact in any case.

**Comment by beckermr** [comment]
Thanks Min. We had access to those for a long time now. Agreed they are not really for our scale.

**Comment by jakirkham** [comment]
Has there been any word from Travis CI on this issue?

**Comment by mfansler** [comment]
Nothing through my email. I am also unable to view [the ticket](https://support.travis-ci.com/hc/requests/56741) they created (always ask for "Sign-in" then dumps me on the Dashboard). Maybe someone from Core should take over.

**Comment by jakirkham** [comment]
Thanks Mervin! 🙏

Have we seen any Travis CI builds run on `linux-ppc64le` (including non-R ones)?

**Comment by mfansler** [comment]
Just pinged Support again and they replied promptly that the issue is still active/visible on their and are working on it.

> *"We are actively investigating this issue. It seems we are getting network related timeouts and we're still troubleshooting on our side about this."*

**Comment by hmaarrfk** [comment]
maybe we can default to Cross compilation + Azure and reduce our system usage there to be nice?

**Comment by jakirkham** [comment]
Yes I made this suggestion here: https://github.com/regro/cf-scripts/issues/2930

Edit: On the R side, think Mervin has been using emulation on Azure. Though think there is some work to look at cross-compilation

**Comment by h-vetinari** [comment]
> On the R side, think Mervin has been using emulation on Azure. Though think there is some work to look at cross-compilation

He [commented](https://github.com/conda-forge/status/issues/185#issuecomment-2277346486) further upthread how to cross-compile R recipes.

**Comment by h-vetinari** [comment]
> > You mean for R or in general?
> 
> In general. The actual feedstock where I'm hitting this is a `go` recipe.

For C/C++ recipes, there's not much to do except change `conda-forge.yml` and rerender. The compilers populated by `{{ compiler("c") }}` etc. will automatically get the right target, with the right activation. For CMake builds, it's good to pass `$CMAKE_ARGS` to the first CMake call, because that contains a bunch of relevant configuration.

The main problem in cross-compilation is that you cannot simply run things (e.g. just-built utilities) during the build process, because the architecture you're compiling for doesn't match what you're running on. That's also why you need the respective dependencies in the `build:` environment (with `# [build_platform != target_platform]`). That's where the cross-python and cross-r stuff come in. For python builds, often numpy/cython etc. are necessary, but basically anything else that's needed for stuff to run _at build time_.

Rust recipes seem to cross-compile without much complications (from a few I've looked at recently), but I'm not familiar with what's necessary for go recipes. The org-wide github-[search](https://github.com/search?q=org%3Aconda-forge+%2F%28%3Fm%29build_platform%3A%5Cn%28%5Cs*.*%5Cn%29*%5Cs*linux_ppc64le%2F+path%3Aconda-forge.yml&type=code) is very useful for finding this sort of thing though. First impression is that you'll have to pay attention to [`GOARCH`](https://github.com/search?q=org%3Aconda-forge+%2FGOARCH%2F+path%3A*.sh&type=code)

**Comment by mfansler** [comment]
Yes, I've mostly been avoiding emulation except in a few edge case that would require heavier patching. The issue in R packages is that the R build process can sometimes involve loading the built library (e.g., to render help). In such cases I'll emulate.

**Comment by h-vetinari** [comment]
Seeing aarch builds fail as well now: https://github.com/conda-forge/povray-feedstock/pull/19

**Comment by jaimergp** [comment]
Should we edit the title and issue description to reflect this new information?

**Comment by mfansler** [comment]
@jaimergp there are still other **linux-aarch64** jobs passing - I'm not convinced that wasn't a sporadic failure. But if non-R feedstocks are seeing consistent failures, the issue description could be generalized.

**Comment by mfansler** [comment]
Travis CI [reports to have resolved the issue](https://www.traviscistatus.com/incidents/12xg3gp8r2gl) and I have confirmed with several jobs that **linux-ppc64le** runs are indeed running normally again.

**Comment by jaimergp** [comment]
Sounds like we can close this soon, then? Let's keep it open for a few more hours just in case, but will close by EOD if we can confirm it's working.

**Comment by jaimergp** [comment]
Checked https://app.travis-ci.com/github/conda-forge and there are several feedstocks with passing builds for both PPC and ARM from few hours ago (e.g. https://app.travis-ci.com/github/conda-forge/databricks-cli-feedstock/builds/272058545?serverType=git). I'll close. Thanks for keeping an eye on this @mfansler!

**Comment by jakirkham** [comment]
Glad this is improving! 🥳 

That said, did just see [a new instance of this]( https://app.travis-ci.com/github/conda-forge/nvtx-c-feedstock/jobs/625551224 )

So doesn't seem like this is fully resolved yet

**Comment by h-vetinari** [comment]
Looking at the travis dashboard, this still seems to be happening to ~50% of PPC jobs (which just get cancelled).

**Comment by h-vetinari** [comment]
At least that aspect can be cured by restarting the job though.

**Comment by mfansler** [comment]
Yeah, looks like it's back to the previous baseline with something like 10%-25% sporadic failure.

**Comment by jaimergp** [comment]
A month has passed and this incident is still open with no foreseeable solution. Are we still observing the 10-25% sporadic failure rate? If that's the case, is it worth studying the feasibility of disabling that platform on Travis CI and let people cross-compile or emulate?

**Comment by jakirkham** [comment]
Was about to ask the same thing. Did see [a build here]( https://app.travis-ci.com/github/conda-forge/datafusion-feedstock/builds/272437497 ). It stalled out in the midst of the build, which is a different issue than this one, but it is an issue that we have seen with Travis CI before

**Comment by mfansler** [comment]
R migration bottlenecked on `xorg-*` migrations for a few weeks (now resolved), so there hasn't been a steady stream of run data for me to estimate recent Travis CI failures rates. However, I'm fine with having this closed, as the acute issue that prompted this report appears over. Should the broader discussion of dropping Travis CI be moved to a dedicated Issue? (where?)

**Comment by mfansler** [comment]
Having more data these last few weeks, I still see high failure rates (~10%-25%). I've updated the OP notification to reflect that more specifically - rather than the acute ppc64le issue - and added notice to consider moving to Azure.

**Comment by jaimergp** [comment]
Looks like we are having issues again, as per this thread: https://conda-forge.zulipchat.com/#narrow/channel/457337-general/topic/travis.20ci.20not.20even.20attaching.20to.20pr

**Comment by martin-g** [comment]
Maybe it is just me but TravisCI almost always fails for the feestocks I care about ...
Usually I use cross-compilation but few times it did not work because the software uses `configure` tests/checks: the code snippet is compiled for the `target_platform` but then executed on the `host_platform` and the error is something like:
```
checking for gethostname in -lnsl... no
checking for gethostbyname in -lnsl... no
checking for accept in -lsocket... no
configure: error: cannot run test program while cross compiling
See `config.log' for more details.
Traceback (most recent call last):
```
Full logs: https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1131643&view=logs&j=d6b58996-039f-5e48-56bf-c3a016e5cd7f&t=dd6a23d9-2356-5264-64bf-579875a5d8d6
https://github.com/conda-forge/ossuuid-feedstock/pull/9 - the OSX ARM64 build. The aarch64/ppc64le fail the same so I dropped cross-compilation for them but now TravisCI is not available at all: 
-  https://github.com/conda-forge/ossuuid-feedstock/pull/9
-  https://github.com/conda-forge/perl-mime-base64-feedstock/pull/2

**Comment by mfansler** [comment]
@martin-g note that Linux ARMs always have the option of emulation on Azure - it's just much slower (e.g., ~5X longer last I check), but typically gets around having to modify things. https://github.com/conda-forge/status/issues/185#issuecomment-2277366715

I think the default `macos-latest` runners on Azure are `macos-14` running on Apple Silicon - we're just not using them yet. Not sure where the discussion stands on rolling that out here. Possibly mixed up with the question of when CF stops supporting `osx-64`.

**Comment by martin-g** [comment]
Thanks! I was not aware of the emulation possibility!
I will try it for https://github.com/conda-forge/ossuuid-feedstock/pull/9 ! Slow is still better than nothing!

**Comment by theAeon** [comment]
cross-referencing this with https://github.com/conda-forge/conda-forge-pinning-feedstock/issues/6595

**Comment by mfansler** [comment]
Given that `conda-smithy` now defaults to `azure` (https://github.com/conda-forge/conda-smithy/pull/2221), I think we can move on from this issue and free up some Status dashboard real estate - closing.
